src/runtime/internal/block_storage.h

```diff
@@ -0,0 +1,366 @@
+#ifndef HALIDE_RUNTIME_BLOCK_STORAGE_H
+#define HALIDE_RUNTIME_BLOCK_STORAGE_H
+
+#include "memory_resources.h"
+
+namespace Halide {
+namespace Runtime {
+namespace Internal {
+
+// Dynamically resizable array for block storage (eg plain old data)
+// -- No usage of constructors/destructors for value type (T)
+// -- Implementation uses memcpy/memmove for copying
+// -- Customizable allocator ... default uses HalideSystemAllocator
+template<typename T>
```

Comment:
We'll need to carefully examine how the use of shared template classes like this affects codegen -- in particular, we should verify that we don't end up with redundant copies of the template methods instantiated into every runtime module (and not deduped by the linker). Ordinarily I wouldn't worry, but our runtime build/link model is unusual enough that we should probably verify that it's really happening.

Reply:
Understood. If template explosion is a big concern, I could just remove the type safety and make these classes just allocate raw bytes, and leave it up to the client code to do the typecasting.